lms/services/jwt.py

```diff
@@ -51,3 +60,63 @@ def encode_with_secret(cls, payload: dict, secret, lifetime):
         jwt_str = jwt.encode(payload, secret, algorithm="HS256")
 
         return jwt_str
+
+    def decode_lti_token(self, id_token: str) -> dict:
+        """
+        Decode an LTI `id_token` JWT.
+
+        The JWT is validated against its corresponding LTIRegistration public key
+        """
+        if not id_token:
+            return {}
+
+        header = jwt.get_unverified_header(id_token)
+        if not header.get("kid"):
+            LOG.debug("Missing 'kid' value in JWT header")
+            return {}
+
+        unverified_payload = jwt.decode(id_token, options={"verify_signature": False})
+        iss, aud = unverified_payload.get("iss"), unverified_payload.get("aud")
+
+        # Find the registration based on the token's claimed issuer & audience
+        registration = self._registration_service.get(iss, aud)
+        if not registration:
+            LOG.debug("Unknown registration for lti_token. iss:%s aud:%s.", iss, aud)
+            return {}
+
+        try:
+            signing_key = self._get_jwk_client(
+                registration.key_set_url
+            ).get_signing_key_from_jwt(id_token)
+
+            return jwt.decode(
+                id_token, key=signing_key.key, audience=aud, algorithms=["RS256"]
+            )
+        except PyJWTError as err:
+            LOG.debug("Invalid JWT. %s", str(err))
+            return {}
+
+    @staticmethod
+    @lru_cache
+    def _get_jwk_client(jwk_url: str):
+        """
+        Get a PyJWKClient for the given key set URL.
+
+        PyJWKClient maintains a cache of keys it has seen we want to keep
+        the clients around with `lru_cache` in case we can reuse that internal cache
+        """
+        return jwt.PyJWKClient(jwk_url)
```

Comment:
Could this be inlined now? It's only used once above, and the result of this calculation is `reified` in the request parameter below, which will serve to store the result.

Alternatively we could add caching around the `decode_lti_token` instead.

Reply:
`_get_jwk_client` maintains the cache across requests, ie for the lifetime of the process/worker. That's why it's declared as a staticmethod. I'll make this more explicit in the docstring.

Internally PyJWKClient keeps a cache of the keys/kid it seen and it won't make any more HTTP requests to get those.